pkg/secrets/secrets.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package secrets
+
+import (
+	"context"
+	"crypto/rsa"
+	"fmt"
+	"hash/crc32"
+
+	secretmanager "cloud.google.com/go/secretmanager/apiv1"
+	"cloud.google.com/go/secretmanager/apiv1/secretmanagerpb"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+)
+
+// GetSecret reads a secret from Secret Manager and validates that it was not
+// corrupted during retrieval. The secretResourceName should be in the format:
+// 'projects/*/secrets/*/versions/*'.
+func GetSecret(ctx context.Context, secretResourceName string) (string, error) {
+	sm, err := secretmanager.NewClient(ctx)
```

Comment:
It's _extremely_ expensive to create this client for each secret. The SecretManager client (like all clients) should be instantiated once and used repeatedly. It should also be closed during shutdown.

Reply:
There are two main scenarios I can think of:
1. Your program needs to access many secrets during the lifetime of its execution. In this case, I agree with you that a single client should be instantiated and re-used and only closed on shutdown. This is what the `AccessSecret` function in this file was meant to address where the secret manager client is passed in as a parameter.
2. Your program only needs to access one and only one secret during the lifetime of its execution. In this case, it makes sense to instantiate the client, fetch the secrete, then close the client since you no longer need it. This is use case the `GetSecret` function was meant to address.

Hopefully any users of these functions would be aware enough of their own programs needs to determine if `GetSecret` or `AccessSecret` should be used.

Reply:
The `Get` versus `Access` semantic is extremely confusing because it diverges from the Secret Manager nomenclature. Get == Read == metadata. Access == secret material.

Reply:
I was unaware of the `Get == Read == metadata` and `Access == secret material` distinction with regards to using secret manager. I've renamed the function to `AccessSecretFromSecretManager` to better indicate that we are actually getting the secret material. I've also extended the function doc to note that the `AccessSecret` function should be used if multiple secrets need to be fetched.